### PR TITLE
Merge TigerBeetle clients into main repo

### PR DESCRIPTION
**Pull request by cb22**
This PR pulls in tigerbeetle-{node,go,java,dotnet}, as well as their respective CI workflows, and adds some extra tooling around publishing to npm and the Go client. 

All commits in this PR - except for the single commit titled `clients: Import clients and scope CI workflows into main repo` - are said imports and can mostly be skipped for review purposes.

- All client commits have been preserved and rewritten to start with `clients/{client}: `. These clients now live under the `src/clients/` subfolder
- Binaries have been removed from the history of the Go and Node clients

For Go:
- To keep binaries out of the main repo, the Go client workflow builds the binaries needed and then pushes the result to the `tigerbeetle-go` repo. It includes the source commit in its commit message, as well as creates a tag called `tigerbeetle-{commit-in-this-repo}` for easy reference.

For Node:
- When working on a PR, the Node workflow will check if the version specified in `package.json` exists on npm. If it does, the hash of the npm hosted package is then compared to a locally built one. 
    - It would be nice to only include / rely on the files that the client _actually_ uses and builds in. At the moment, the detection for this will likely have some false positives.
- If the hashes aren't identical, it'll throw an error. The idea is to stop cases where changes have been made to the client, but a new release hasn't been cut (thanks @sentientwaffle!)
- When pushing to main or merging, a new release can be created by simply adjusting the version number in `package.json`

Before merging:
- [x] The `tigerbeetle-go` repo needs to have a deploy key with RW access added, and that key added as a secret here under `TIGERBEETLE_GO_DEPLOY_KEY`
- [x] A key for npm publishing needs to be added as a secret under `TIGERBEETLE_NODE_PUBLISH_KEY`

Other points:
- Currently, the workflows for the clients are not scoped to the client path; they'll trigger if something unrelated is changed. While changing the `dotnet` client will very likely have 0 bearing on the `java`, this is useful to catch modifications and build for things like the core Zig client library that they pull in.

### CODE REVIEW

File: .github/workflows/clients-go-build.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+        uses: actions/cache@v3
+        env:
+          cache-name: cache-go-binaries
+        with:
+          path: src/clients/go/pkg/native
+          key: go-binaries-${{ github.sha }}
+          restore-keys: |
+            go-binaries-${{ github.sha }}
+      - run: mkdir ~/.ssh && echo -e "${{ secrets.TIGERBEETLE_GO_DEPLOY_KEY }}" > ~/.ssh/id_ed25519 && chmod 600 ~/.ssh/id_ed25519
+      - run: git clone git@github.com:tigerbeetledb/tigerbeetle-go.git ~/tigerbeetle-go && cp -R . ~/tigerbeetle-go
```
Comment (sentientwaffle) [inline .github/workflows/clients-go-build.yml]: If a file is moved in `tigerbeetledb/tigerbeetle`, will this leave the old copy intact in `tigerbeetledb/tigerbeetle-go`? Maybe we need to delete the files (except `.git`) in `~/tigerbeetle-go` before copying `./src/clients/go/*` in.

File: .github/workflows/clients-node-build.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  version_check:
+    needs: [shellcheck_scripts, test_distro_install]
+    runs-on: ubuntu-latest
+    environment: main
+    steps:
+      - uses: actions/checkout@v3
+      - run: npm install && ./scripts/version_check.sh
+
+  publish:
+    if: ${{ github.event_name == 'push' && github.ref == 'refs/heads/main' }}
```
Comment (sentientwaffle) [inline .github/workflows/clients-node-build.yml:44]: Just to check my understanding, to publish a new version, we just need to:

- bump the `version` in `package.json`,
- and then merge the commit that bumps the version into `main`.

File: src/clients/node/package.json [hunk 3]
```diff
@@ -39,7 +40,7 @@
     "build": "npm run build_tsc && npm run build_lib",
     "build_tsc": "./node_modules/typescript/bin/tsc",
     "build_lib": "mkdir -p dist && zig/zig build-lib -mcpu=baseline -OReleaseSafe -dynamic -lc -isystem build/node-$(node --version)/include/node src/node.zig -fallow-shlib-undefined -femit-bin=dist/client.node",
-    "prepack": "git submodule deinit --all && git submodule update --init && npm run build_tsc",
+    "prepack": "npm run build_tsc && rm -rf src/zig-cache",
```
Comment (sentientwaffle) [inline src/clients/node/package.json]: Why is `rm -rf src/zig-cache` needed? My understanding of `npm` (maybe wrong) is that the `files` list (above) enumerates the files packaged.

**Review by sentientwaffle — COMMENTED** [review]
The deploy & publish keys are set up.

**Comment by eatonphil** [comment]
Nice work, Fed!

File: .github/workflows/clients-go-build.yml [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  pull_request:
+    branches:
+      - 'main'
+
+jobs:
+  shellcheck_scripts:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: ./scripts/shellcheck.sh
```
Comment (eatonphil) [inline .github/workflows/clients-go-build.yml:24]: We can probably now drop all individual calls to shellcheck and just have one global one?

File: .github/workflows/clients-go-build.yml [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        env:
+          cache-name: cache-go-binaries
+        with:
+          path: src/clients/go/pkg/native
+          key: go-binaries-${{ github.sha }}
+          restore-keys: |
+            go-binaries-${{ github.sha }}
+      - run: go test ./...
+
+  # TODO: Windows exit codes broken
```
Comment (eatonphil) [inline .github/workflows/clients-go-build.yml:62]:
```
Set-StrictMode -Version 3
$ErrorActionPreference = "Stop"
$PSDefaultParameterValues['*:ErrorAction']='Stop'
$LASTEXITCODE = 0
```

Here's what I do [elsewhere](https://github.com/multiprocessio/datastation/blob/main/scripts/ci/prepare_windows.ps1).

File: .github/workflows/clients-node-build.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  pull_request:
+    branches:
+      - 'main'
+
+jobs:
+  shellcheck_scripts:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: ./scripts/shellcheck.sh
```
Comment (eatonphil) [inline .github/workflows/clients-node-build.yml:24]: This could be dropped since a global check should catch this

**Comment by eatonphil** [comment]
> To keep binaries out of the main repo, the Go client workflow builds the binaries needed and then pushes the result to the tigerbeetle-go repo. It includes the source commit in its commit message, as well as creates a tag called tigerbeetle-{commit-in-this-repo} for easy reference.

I don't understand the benefit of keeping binaries out of the repo.

**Comment by sentientwaffle** [comment]
> I don't understand the benefit of keeping binaries out of the repo.

This was my request — I prefer not to include them b/c they make checkouts larger, and my workflow involves around half a dozen branches checked out simultaneously (via git worktrees) :wink:.

**Comment by eatonphil** [comment]
> This was my request — I prefer not to include them b/c they make checkouts larger, and my workflow involves around half a dozen branches checked out simultaneously (via git worktrees) 😉.

Gotcha.

File: .github/workflows/clients-go-build.yml (hunk 1 above)
Comment (cb22) [reply .github/workflows/clients-go-build.yml]: Good catch - we need to do exactly that, will sort it out.

File: .github/workflows/clients-node-build.yml (hunk 2 above)
Comment (cb22) [reply .github/workflows/clients-node-build.yml:44]: Yep!

File: src/clients/node/package.json (hunk 3 above)
Comment (cb22) [reply src/clients/node/package.json]: It's not anymore, I'll take that out. I originally added it before I modified the `files` to be more specific.

File: .github/workflows/clients-node-build.yml (hunk 2 above)
Comment (sentientwaffle) [reply .github/workflows/clients-node-build.yml:44]: That's perfect; thank you!

File: .github/workflows/clients-go-build.yml (hunk 4 above)
Comment (cb22) [reply .github/workflows/clients-go-build.yml:24]: Yep! I can do this in here, or as a separate commit once this is merged?

File: .github/workflows/clients-go-build.yml (hunk 5 above)
Comment (cb22) [reply .github/workflows/clients-go-build.yml:62]: Cool thanks - I'll get this sorted

File: .github/workflows/clients-go-build.yml (hunk 4 above)
Comment (eatonphil) [reply .github/workflows/clients-go-build.yml:24]: Whatever you'd like

Comment (sentientwaffle) [reply .github/workflows/clients-go-build.yml:24]: Either is fine; whichever is simpler for you.

**Comment by cb22** [comment]
I've pushed a commit addressing some of the issues above - the Go repo files & npm prepack.

The `.bat` Windows tests and shell-check I'll do in a separate PR, just to keep things more manageable!

We'll also need to enable some more actions in our project settings:

- `actions/checkout@v3`
- `actions/cache@v3`
- `actions/setup-java@v3`
- (or perhaps `actions/*`?)

**Comment by sentientwaffle** [comment]
I added
```
actions/checkout@v3
actions/cache@v3
actions/setup-java@v3
```
to our allowed-actions list.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**